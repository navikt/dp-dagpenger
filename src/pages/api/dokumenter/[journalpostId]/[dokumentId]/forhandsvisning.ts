import { logger } from "@navikt/next-logger";
import { NextApiHandler } from "next";
import { v4 as uuidv4 } from "uuid";
import { getSession } from "../../../../../lib/auth.utils";

const audience = `${process.env.SAF_SELVBETJENING_CLUSTER}:teamdokumenthandtering:${process.env.SAF_SELVBETJENING_SCOPE}`;

export const config = {
  api: {
    responseLimit: false,
  },
};

async function hentDokument(
  token: string,
  journalpostId: string,
  dokumentInfoId: string,
  callId: uuidv4,
): Promise<Response> {
  const endpoint = `${process.env.SAF_SELVBETJENING_INGRESS}/rest/hentdokument/${journalpostId}/${dokumentInfoId}/ARKIV`;

  const headers = {
    Authorization: `Bearer ${token}`,
    "Nav-Callid": callId,
    "Nav-Consumer-Id": "dp-dagpenger",
  };

  return fetch(endpoint, { headers, cache: "no-store" });
}

const handleHentDokument: NextApiHandler<Buffer> = async (req, res) => {
  const session = await getSession(req);
  if (!session.token) return res.status(401).end();

  const { journalpostId, dokumentId } = req.query;
  const callId = uuidv4();
  return hentDokument(
    await session.apiToken(audience),
    <string>journalpostId,
    <string>dokumentId,
    callId,
  )
    .then(async (dokumentResponse) => {
      res.setHeader(
        "Content-Disposition",
        dokumentResponse.headers.get("Content-Disposition") || "inline",
      );

      res.setHeader(
        "Content-Type",
        dokumentResponse.headers.get("Content-Type") || "application/octet-stream",
      );

      // Vi kan ikke bruke ReadableStream direkte fra fetch her, går over til å returnere buffer i stedet.
      // Vi fant feil med ReadableStream.pipe når vi oppgraderte til Node v18, som har innebygget fetch
      // (fra web). Før hadde de polyfillet en fetch fra node.
      // Senere kan vi teste ut https://www.npmjs.com/package/readable-web-to-node-stream
      const arrayBuffer = await dokumentResponse.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);

      return res.status(dokumentResponse.status).send(buffer);
    })
    .catch((errors) => {
      console.log(errors);
      logger.error(`Feil fra SAF med call-id ${callId}: ${errors}`);
      return res.status(500).send(errors);
    });
};

export default handleHentDokument;
