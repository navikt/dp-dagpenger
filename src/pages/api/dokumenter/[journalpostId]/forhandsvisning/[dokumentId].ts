import { NextApiResponse } from "next";
import { v4 as uuidv4 } from "uuid";
import {
  AuthedNextApiRequest,
  withMiddleware,
} from "../../../../../auth/middlewares";

const audience = `${process.env.SAF_SELVBETJENING_CLUSTER}:teamdokumenthandtering:safselvbetjening`;

async function hentDokument(
  token: string,
  journalpostId: string,
  dokumentInfoId: string
): Promise<Blob> {
  const callId = uuidv4();
  const endpoint = `${process.env.SAF_SELVBETJENING_INGRESS}/rest/hentdokument/${journalpostId}/${dokumentInfoId}/ARKIV`;

  const headers = {
    Authorization: `Bearer ${token}`,
    "Nav-Callid": callId,
    "Nav-Consumer-Id": "dp-dagpenger",
    Accept: "application/pdf",
  };

  try {
    console.log(`Henter dokument med call-id: ${callId}`);
    return await fetch(endpoint, { headers }).then((res) => res.blob());
  } catch (error) {
    console.error(`Feil fra SAF med call-id ${callId}: ${error}`);
    throw error;
  }
}

export async function handleDokumenter(
  req: AuthedNextApiRequest,
  res: NextApiResponse
) {
  const user = req.user;
  if (!user) return res.status(401).end();
  const token = await user.tokenFor(audience);
  const { journalpostId, dokumentInfoId } = req.query;

  try {
    return res
      .setHeader("Content-type", "application/pdf")
      .send(
        await hentDokument(token, <string>journalpostId, <string>dokumentInfoId)
      );
  } catch (errors) {
    return res.status(500).send(errors);
  }
}

export default withMiddleware(handleDokumenter);
