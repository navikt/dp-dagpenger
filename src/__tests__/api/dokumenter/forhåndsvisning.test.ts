import { createMocks } from "node-mocks-http";
import { handleHentDokument } from "../../../pages/api/dokumenter/[journalpostId]/[dokumentId]/forhandsvisning";

describe("/api/dokumenter/[journalpostId]/[dokumentId]/forhandsvisning", () => {
  test("svarer med en liste dokumenter", async () => {
    const { req, res } = createMocks({
      method: "GET",
      query: {
        journalpostId: 123,
        dokumentId: 345,
      },
      user: {
        fnr: "123123123",
        tokenset: {
          access_token: "123",
        },
        tokenFor: (token) => token,
      },
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await handleHentDokument(req, res);

    expect(res._getStatusCode()).toBe(200);
    expect(res._getData()).toMatchSnapshot();
  });
});
