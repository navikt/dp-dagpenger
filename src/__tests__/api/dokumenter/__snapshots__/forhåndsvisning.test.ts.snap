// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/api/dokumenter/[journalpostId]/[dokumentId]/forhandsvisning > svarer med en liste dokumenter 1`] = `""`;
